pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/RoundRobinScheduler.java

```diff
@@ -39,168 +47,208 @@
-  // while the _ready queue contains the operator chains that are ready
-  // to be scheduled (have data, or are first-time scheduled)
-  private final Queue<AvailableEntry> _available = new LinkedList<>();
-  private final Queue<OpChain> _ready = new LinkedList<>();
-
-  private boolean _isShutDown = false;
-
-  // using a Set here is acceptable because calling hasNext() and
-  // onDataAvailable() cannot occur concurrently - that means that
-  // anytime we schedule a new operator based on the presence of
-  // mail we can be certain that it will consume all of the mail
-  // form that mailbox, even if there are multiple items in it. If,
-  // during execution of that operator, more mail appears, then the
-  // operator will be rescheduled immediately potentially resulting
-  // in a false-positive schedule
-  @VisibleForTesting
-  final Set<MailboxIdentifier> _seenMail = new HashSet<>();
+  private final Map<OpChainId, OpChain> _aliveChains = new ConcurrentHashMap<>();
+  final Set<OpChainId> _seenMail = Sets.newConcurrentHashSet();
+  private final Map<OpChainId, Long> _available = new ConcurrentHashMap<>();
 
-  public RoundRobinScheduler() {
-    this(DEFAULT_RELEASE_TIMEOUT);
-  }
+  private final BlockingQueue<OpChain> _ready = new LinkedBlockingQueue<>();
+
+  private final Lock _lock = new ReentrantLock();
+  private final ScheduledExecutorService _availableOpChainReleaseService;
 
-  public RoundRobinScheduler(long releaseTimeout) {
-    this(releaseTimeout, System::currentTimeMillis);
+  public RoundRobinScheduler(long releaseTimeoutMs) {
+    this(releaseTimeoutMs, System::currentTimeMillis);
   }
 
-  public RoundRobinScheduler(long releaseTimeoutMs, Supplier<Long> ticker) {
-    _releaseTimeout = releaseTimeoutMs;
+  RoundRobinScheduler(long releaseTimeoutMs, Supplier<Long> ticker) {
+    Preconditions.checkArgument(releaseTimeoutMs > 0, "Release timeout for round-robin scheduler should be > 0ms");
+    _releaseTimeoutMs = releaseTimeoutMs;
     _ticker = ticker;
+    _availableOpChainReleaseService = Executors.newSingleThreadScheduledExecutor(r -> {
+      Thread t = new Thread(r);
+      t.setName(AVAILABLE_RELEASE_THREAD_NAME);
+      t.setDaemon(true);
+      return t;
+    });
+    _availableOpChainReleaseService.scheduleAtFixedRate(() -> {
+      List<OpChainId> timedOutWaiting = new ArrayList<>();
+      for (Map.Entry<OpChainId, Long> entry : _available.entrySet()) {
+        if (Thread.interrupted()) {
+          LOGGER.warn("Thread={} interrupted. Scheduler may be shutting down.", AVAILABLE_RELEASE_THREAD_NAME);
+          break;
+        }
+        if (_ticker.get() > entry.getValue()) {
+          timedOutWaiting.add(entry.getKey());
+        }
+      }
+      for (OpChainId opChainId : timedOutWaiting) {
+        _lock.lock();
+        try {
+          if (_available.containsKey(opChainId)) {
+            _available.remove(opChainId);
+            _ready.offer(_aliveChains.get(opChainId));
```

Comment:
this offer is blocked or error out if the queue is full. but I guess this is currently not a concern?